File: collector/providers/cursor.py
"""Provedor Cursor: JWT local (state.vscdb) + Dashboard Connect RPC / auth/usage legado."""

from __future__ import annotations

from typing import Any

from cursor_state import cursor_token_and_plan, cursor_token_candidates, jwt_expired
from formatting import as_percent, cycle_end_label, money_cents
from http_util import http_json

CURSOR_USAGE_URL = (
    "https://api2.cursor.sh/aiserver.v1.DashboardService/GetCurrentPeriodUsage"
)
CURSOR_AUTH_USAGE_URL = "https://api2.cursor.sh/auth/usage"


def parse_cursor_dashboard(data: dict[str, Any], plan: str | None) -> dict[str, Any] | None:
    usage = data.get("planUsage") or data.get("plan_usage") or {}
    if not isinstance(usage, dict):
        usage = {}
    # Painel Cursor: "Cursor Models" = auto, "Other Models" = api.
    percent = as_percent(usage.get("autoPercentUsed") or usage.get("totalPercentUsed"))
    other_percent = as_percent(usage.get("apiPercentUsed"))
    spend = data.get("spendLimitUsage") or data.get("spend_limit_usage") or {}
    if not isinstance(spend, dict):
        spend = {}
    ondemand_limit = money_cents(spend.get("individualLimit") or spend.get("limit"))
    ondemand_remain = money_cents(spend.get("individualRemaining") or spend.get("remaining"))
    ondemand_used = None
    if ondemand_limit is not None and ondemand_remain is not None:
        ondemand_used = max(0, ondemand_limit - ondemand_remain)
    elif ondemand_limit is not None:
        ondemand_used = 0
    cycle_end = cycle_end_label(
        data.get("billingCycleEnd") or data.get("billing_cycle_end") or usage.get("endDate")
    )
    if percent is None and other_percent is None and ondemand_limit is None and not cycle_end:
        return None
    return {
        "ok": percent is not None or other_percent is not None or ondemand_limit is not None,
        "error": None
        if (percent is not None or other_percent is not None or ondemand_limit is not None)
        else "planUsage sem números",
        "percent": percent,
        "other_percent": other_percent,
        "used_cents": ondemand_used,
        "limit_cents": ondemand_limit,
        "remaining_cents": ondemand_remain,
        "bonus_cents": 0,
        "cycle_end": cycle_end,
        "plan": (plan or data.get("membershipType") or "").strip() or None,
        "requests_used": None,
        "requests_limit": None,
    }


def parse_cursor_auth_usage(data: dict[str, Any], plan: str | None) -> dict[str, Any]:
    # Formato legado: { "gpt-4": { "numRequests": n, "maxRequestUsage": m }, ... }
    best_pct = None
    used = limit = None
    if isinstance(data, dict):
        for _key, bucket in data.items():
            if not isinstance(bucket, dict):
                continue
            n = bucket.get("numRequests")
            m = bucket.get("maxRequestUsage") or bucket.get("maxRequests")
            try:
                n_i, m_i = int(n), int(m) if m is not None else 0
            except (TypeError, ValueError):
                continue
            if m_i <= 0:
                continue
            pct = as_percent((n_i / m_i) * 100.0)
            if best_pct is None or (pct is not None and pct > best_pct):
                best_pct, used, limit = pct, n_i, m_i
    ok = best_pct is not None
    return {
        "ok": ok,
        "error": None if ok else "auth/usage sem buckets",
        "percent": best_pct,
        "other_percent": None,
        "used_cents": None,
        "limit_cents": None,
        "remaining_cents": None,
        "bonus_cents": None,
        "cycle_end": None,
        "plan": plan,
        "requests_used": used,
        "requests_limit": limit,
    }


def _cursor_fail(msg: str) -> dict[str, Any]:
    return {
        "ok": False,
        "error": msg,
        "percent": None,
        "other_percent": None,
        "used_cents": None,
        "limit_cents": None,
        "remaining_cents": None,
        "bonus_cents": None,
        "cycle_end": None,
        "plan": None,
        "requests_used": None,
        "requests_limit": None,
    }


def _is_auth_error(msg: str) -> bool:
    return "HTTP 401" in msg or "HTTP 403" in msg


def _fetch_cursor_with_token(token: str, plan: str | None) -> dict[str, Any]:
    headers = {
        "Authorization": f"Bearer {token}",
        "Content-Type": "application/json",
        "Connect-Protocol-Version": "1",
        "Accept": "application/json",
    }
    dash_err = "GetCurrentPeriodUsage vazio"
    try:
        data = http_json(
            CURSOR_USAGE_URL,
            method="POST",
            headers=headers,
            body=b"{}",
        )
        if isinstance(data, dict):
            parsed = parse_cursor_dashboard(data, plan)
            if parsed and parsed.get("ok"):
                return parsed
            if parsed and parsed.get("error"):
                dash_err = str(parsed["error"])
    except RuntimeError as exc:
        dash_err = str(exc)

    try:
        data = http_json(
            CURSOR_AUTH_USAGE_URL,
            headers={"Authorization": f"Bearer {token}", "Accept": "application/json"},
        )
        if isinstance(data, dict):
            parsed = parse_cursor_auth_usage(data, plan)
            if parsed.get("ok"):
                return parsed
            return _cursor_fail(f"{dash_err}; {parsed.get('error')}")
    except RuntimeError as exc:
        return _cursor_fail(f"{dash_err}; fallback: {exc}")
    return _cursor_fail(dash_err)


def fetch_cursor() -> dict[str, Any]:
    cands = cursor_token_candidates()
    if not cands:
        _token, _plan, err = cursor_token_and_plan()
        return _cursor_fail(err or "sem JWT Cursor")
    last_err = "sem JWT Cursor"
    for _source, token, plan in cands:
        if jwt_expired(token):
            last_err = "JWT expirado; abra o Cursor neste Mac"
            continue
        parsed = _fetch_cursor_with_token(token, plan)
        if parsed.get("ok"):
            return parsed
        err = str(parsed.get("error") or "")
        last_err = err
        if _is_auth_error(err):
            continue
        return parsed
    return _cursor_fail(last_err)

File: backend/src/providers/cursor.ts
/**
 * Provedor Cursor: JWT local + contas extras via Dashboard Connect RPC.
 * Port of backend-python-legacy/app/providers/cursor.py
 */
import { asPercentPoints, isoOrNone, moneyCents, pick, ratioPercent } from "../formatting.js";
import { httpJson } from "../httpClient.js";
import { cursorTokenCandidates, jwtExpired } from "../local/cursorState.js";
import { provider as providerCfg } from "../store.js";

export const CURSOR_USAGE_URL = "https://api2.cursor.sh/aiserver.v1.DashboardService/GetCurrentPeriodUsage";
export const CURSOR_AUTH_USAGE_URL = "https://api2.cursor.sh/auth/usage";

export function cursorFail(msg: string): Record<string, unknown> {
  return {
    ok: false,
    error: msg,
    percent: null,
    other_percent: null,
    used_cents: null,
    limit_cents: null,
    remaining_cents: null,
    bonus_cents: null,
    cycle_end: null,
    plan: null,
    requests_used: null,
    requests_limit: null,
  };
}

export function parseCursorDashboard(data: Record<string, unknown>, plan: string | null): Record<string, unknown> | null {
  const usageRaw = (data.planUsage ?? data.plan_usage ?? {}) as unknown;
  const usage = usageRaw !== null && typeof usageRaw === "object" && !Array.isArray(usageRaw) ? (usageRaw as Record<string, unknown>) : {};

  // use asPercentPoints — 0 is valid, so use pick not ||
  let percent = asPercentPoints(pick(usage.autoPercentUsed, usage.auto_percent_used));
  let otherPercent = asPercentPoints(pick(usage.apiPercentUsed, usage.api_percent_used));

  const spendRaw = (data.spendLimitUsage ?? data.spend_limit_usage ?? {}) as unknown;
  const spend = spendRaw !== null && typeof spendRaw === "object" && !Array.isArray(spendRaw) ? (spendRaw as Record<string, unknown>) : {};

  const ondemandLimit = moneyCents(pick(spend.individualLimit, spend.limit));
  const ondemandRemain = moneyCents(pick(spend.individualRemaining, spend.remaining));
  let ondemandUsed: number | null = null;
  if (ondemandLimit !== null && ondemandRemain !== null) {
    ondemandUsed = Math.max(0, ondemandLimit - ondemandRemain);
  } else if (ondemandLimit !== null) {
    ondemandUsed = 0;
  }

  const bonus = moneyCents(pick(usage.bonusSpend, usage.bonus_spend));

  const cycleEnd = isoOrNone(
    pick(data.billingCycleEnd, data.billing_cycle_end, usage.endDate, usage.end_date),
  );

  // proto3 omits scalar 0: ciclo novo chega sem autoPercentUsed/apiPercentUsed.
  if (percent === null && otherPercent === null && ondemandLimit === null && !cycleEnd) {
    return null;
  }
  if (percent === null) percent = 0.0;
  if (otherPercent === null) otherPercent = 0.0;

  return {
    ok: true,
    error: null,
    percent,
    other_percent: otherPercent,
    used_cents: ondemandUsed,
    limit_cents: ondemandLimit,
    remaining_cents: ondemandRemain,
    bonus_cents: bonus !== null ? bonus : 0,
    cycle_end: cycleEnd,
    plan: (plan ?? String(data.membershipType ?? "").trim()) || null,
    requests_used: null,
    requests_limit: null,
  };
}

export function parseCursorAuthUsage(data: Record<string, unknown>, plan: string | null): Record<string, unknown> {
  let bestPct: number | null = null;
  let used: number | null = null;
  let limit: number | null = null;

  for (const bucket of Object.values(data)) {
    if (bucket === null || typeof bucket !== "object" || Array.isArray(bucket)) continue;
    const b = bucket as Record<string, unknown>;
    const n = b.numRequests;
    const m = b.maxRequestUsage ?? b.maxRequests;
    let nI: number, mI: number;
    try {
      nI = Math.trunc(Number(n));
      mI = m !== null && m !== undefined ? Math.trunc(Number(m)) : 0;
      if (Number.isNaN(nI) || Number.isNaN(mI)) continue;
    } catch {
      continue;
    }
    if (mI <= 0) continue;
    const pct = ratioPercent(nI, mI);
    if (bestPct === null || (pct !== null && pct > bestPct)) {
      bestPct = pct;
      used = nI;
      limit = mI;
    }
  }

  const ok = bestPct !== null;
  return {
    ok,
    error: ok ? null : "auth/usage sem buckets",
    percent: bestPct,
    other_percent: null,
    used_cents: null,
    limit_cents: null,
    remaining_cents: null,
    bonus_cents: null,
    cycle_end: null,
    plan,
    requests_used: used,
    requests_limit: limit,
  };
}

function isAuthError(msg: string): boolean {
  return msg.includes("HTTP 401") || msg.includes("HTTP 403");
}

async function fetchCursorWithToken(token: string, plan: string | null): Promise<Record<string, unknown>> {
  const headers: Record<string, string> = {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
    "Connect-Protocol-Version": "1",
    Accept: "application/json",
  };
  let dashErr = "GetCurrentPeriodUsage vazio";
  try {
    const data = await httpJson(CURSOR_USAGE_URL, {
      method: "POST",
      headers,
      body: Buffer.from("{}"),
      provider: "CURSOR",
    });
    if (data !== null && typeof data === "object" && !Array.isArray(data)) {
      const parsed = parseCursorDashboard(data as Record<string, unknown>, plan);
      if (parsed && parsed.ok) return parsed;
      if (parsed && parsed.error) dashErr = String(parsed.error);
    }
  } catch (e) {
    dashErr = String(e);
  }

  try {
    const data = await httpJson(CURSOR_AUTH_USAGE_URL, {
      headers: { Authorization: `Bearer ${token}`, Accept: "application/json" },
      provider: "CURSOR",
    });
    if (data !== null && typeof data === "object" && !Array.isArray(data)) {
      const parsed = parseCursorAuthUsage(data as Record<string, unknown>, plan);
      if (parsed.ok) return parsed;
      return cursorFail(`${dashErr}; ${parsed.error}`);
    }
  } catch (e) {
    return cursorFail(`${dashErr}; fallback: ${e}`);
  }
  return cursorFail(dashErr);
}

export async function fetchCursorOne(token: string, plan: string | null): Promise<Record<string, unknown>> {
  token = (token ?? "").trim();
  if (!token) return cursorFail("sem token Cursor");
  return fetchCursorWithToken(token, plan);
}

async function fetchCursorLocal(cands: Array<[string, string, string | null]>): Promise<Record<string, unknown>> {
  let lastErr = "sem JWT Cursor";
  for (const [, token, plan] of cands) {
    if (jwtExpired(token)) {
      lastErr = "JWT expirado; abra o Cursor neste computador";
      continue;
    }
    const parsed = await fetchCursorOne(token, plan);
    if (parsed.ok) return parsed;
    const err = String(parsed.error ?? "");
    lastErr = err;
    if (isAuthError(err)) continue;
    return parsed;
  }
  return cursorFail(lastErr);
}

export async function fetchCursorAccounts(cfg: Record<string, unknown>): Promise<Array<Record<string, unknown>>> {
  const accounts: Array<Record<string, unknown>> = [];
  const p = providerCfg(cfg, "cursor") as Record<string, unknown>;

  if (!p.hidden) {
    const localCands = cursorTokenCandidates(cfg);
    if (localCands.length > 0) {
      const result = await fetchCursorLocal(localCands);
      accounts.push({ id: "local", label: String(p.local_label ?? ""), ...result });
    }
  }

  let extra = Array.isArray(p.accounts) ? [...(p.accounts as unknown[])] : [];
  if (extra.length === 0) {
    const legacy = String(p.paste_secret ?? "").trim();
    if (legacy) extra = [{ id: "legacy", label: "", secret: legacy }];
  }
  for (const accRaw of extra) {
    const acc = accRaw as Record<string, unknown>;
    const token = String(acc.secret ?? "").trim();
    const label = String(acc.label ?? "").trim();
    const aid = String(acc.id ?? "extra");
    const result = await fetchCursorOne(token, null);
    accounts.push({ id: aid, label, ...result });
  }
  return accounts;
}

export const cursor_fail = cursorFail;
export const parse_cursor_dashboard = parseCursorDashboard;
export const parse_cursor_auth_usage = parseCursorAuthUsage;
export const fetch_cursor_one = fetchCursorOne;
export const fetch_cursor_accounts = fetchCursorAccounts;
